src/log_rate.c

```diff
@@ -0,0 +1,110 @@
+static void write_io_bufs(stdpipe_t pipe, char* buf, ssize_t count);
+static void sleep_for_the_rest_of_this_period();
+static void start_new_period();
+
+void log_rate_init() {
+	start_new_period();
+}
+
+bool log_rate_write_to_logs(stdpipe_t pipe, char *buf, ssize_t num_read) {
+	char* buf_start = buf;
+	ssize_t bytes_remaining = num_read;
+	struct timespec now;
+	clock_gettime(CLOCK_MONOTONIC, &now);
+	int64_t diff_nano = subtract_timespecs_nano(&now, &start_of_this_period);
+	if (diff_nano < SECS_PER_PERIOD * BILLION) {
+		ssize_t bytes_we_can_write = BYTES_PER_PERIOD - bytes_written_this_period;
+		if (num_read <= bytes_we_can_write) {
+			write_io_bufs(pipe, buf_start, num_read);
+			bytes_written_this_period += num_read;
+			return true;
+		} else {
+			write_io_bufs(pipe, buf_start, bytes_we_can_write);
+			bytes_written_this_period += bytes_we_can_write;
+			buf_start += bytes_we_can_write;
+			bytes_remaining = num_read - bytes_we_can_write;
+			sleep_for_the_rest_of_this_period();
+			start_new_period();
+		}
+	} else {
+		start_new_period();
+	}
+
+	ssize_t chunks = bytes_remaining / BYTES_PER_PERIOD;
+	ssize_t remainder = bytes_remaining % BYTES_PER_PERIOD;
+
+	for (ssize_t i = 0; i < chunks; ++i) {
+		write_io_bufs(pipe, buf_start + i * BYTES_PER_PERIOD, BYTES_PER_PERIOD);
+		sleep_for_the_rest_of_this_period();
+		start_new_period();
+	}
+	if (remainder != 0) {
+	    if (bytes_written_this_period + remainder > BYTES_PER_PERIOD) {
+		    sleep_for_the_rest_of_this_period();
+		    start_new_period();
+	    }
+	    write_io_bufs(pipe, buf_start + (chunks * BYTES_PER_PERIOD), remainder);
+	    bytes_written_this_period += remainder;
+	}
+	return true;
+}
+
+struct timespec subtract_timespecs(const struct timespec* first, const struct timespec* second) {
+	int64_t diff_nanoseconds = subtract_timespecs_nano(first, second);
+	struct timespec ret = {diff_nanoseconds / BILLION, diff_nanoseconds % BILLION};
+	return ret;
+}
+
+int64_t subtract_timespecs_nano(const struct timespec* first, const struct timespec* second) {
+	return (first->tv_sec - second->tv_sec) * BILLION + first->tv_nsec - second->tv_nsec;
+}
+
+void write_io_bufs(stdpipe_t pipe, char* buf, ssize_t count) {
+	ssize_t chunks = count / IO_BUF_SIZE;
+	ssize_t remainder = count % IO_BUF_SIZE;
+	for (int i = 0; i < chunks; ++i) {
+		write_to_logs(pipe, buf + (i * IO_BUF_SIZE), IO_BUF_SIZE);
+	}
+	write_to_logs(pipe, buf + count - remainder, remainder);
+}
+
+void sleep_for_the_rest_of_this_period() {
+	int ret;
+	struct timespec now, sleep, diff;
+	clock_gettime(CLOCK_MONOTONIC, &now);
+	diff = subtract_timespecs(&now, &start_of_this_period);
+	sleep = subtract_timespecs(&secs_per_period, &diff);
+	if (sleep.tv_sec < 0 || sleep.tv_nsec < 0)
+		nexit("negative sleep");
+	do {
+		ret = nanosleep(&sleep, &sleep);
+	} while (ret == -1 && errno == EINTR);
+}
+
+void start_new_period() {
+	bytes_written_this_period = 0;
+	clock_gettime(CLOCK_MONOTONIC, &start_of_this_period);
```

Comment:
Isn't this going to cause the periods to drift?  Instead, could we just use math to calculate the periods from some initial startup time?

Reply:
Could you illustrate with a bit of code how you propose to calculate the periods?

Reply:
Cautionary note: clock_gettime() is relatively expensive in tight inner loops. I'm not saying this code is calling it needlessly, but keep in mind as you fiddle with the timing calculations. A common strategy is to cache the time for re-use and only refresh it after blocking IO calls.

Reply:
I think this code is pretty close to that idea. It finds out the time the first thing on entering this code and then the usual sequence is
write()
clock_gettime() // how long to sleep for?
nanosleep() // sleep
clock_gettime() // when does the new period start
Not sure if it's worth/good idea eliminating the last clock_gettime() and just adding the sleep time to the previous timestamp. Seems like a topic for a profiling session once we get to that point.